src/mol-plugin-state/manager/focus-camera/focus-first-residue.ts

```diff
@@ -0,0 +1,162 @@
+        const firstChainPositions = [];
+        if (AtomIndexs) {
+            for (let i = 0; i < AtomIndexs.length; i++) {
+                const coordinates = structure.units[0].conformation.position(AtomIndexs[i], tmpMatrixPos);
+                for (let j = 0; j < coordinates.length; j++) {
+                    firstChainPositions.push(coordinates[j]);
+                }
+            }
+            let sumX = 0;
+            let sumY = 0;
+            let sumZ = 0;
+            for (let i = 0; i < firstChainPositions.length; i += 3) {
+                sumX += firstChainPositions[i];
+                sumY += firstChainPositions[i + 1];
+                sumZ += firstChainPositions[i + 2];
+            }
+            const averagePosition = Vec3.zero();
+            averagePosition[0] = sumX / AtomIndexs.length;
+            averagePosition[1] = sumY / AtomIndexs.length;
+            averagePosition[2] = sumZ / AtomIndexs.length;
+            return averagePosition;
+        } else {
+            return Vec3.create(caPositions[0], caPositions[1], caPositions[2]);
+        }
+    }
+
+}
+
+export function pcaFocus(plugin: PluginContext, options: Partial<CameraFocusOptions> & { principalAxes?: PrincipalAxes }, positionToFlip?: Vec3) {
+    if (options?.principalAxes) {
+        const { origin, dirA, dirB, dirC } = options.principalAxes.boxAxes;
+        let toFlip: String[] = [];
+        if (positionToFlip) {
+            toFlip = getAxesToFlip(positionToFlip, origin, dirA, dirB);
+        }
+        toFlip.forEach((axis)=>{
+            if (axis === 'aroundY') {
+                Vec3.negate(dirC, dirC);
+            } else if (axis === 'aroundX') {
+                Vec3.negate(dirA, dirA);
+                Vec3.negate(dirC, dirC);
+            }
+        });
+        if (plugin.canvas3d) {
+            plugin.canvas3d.camera.setState(Camera.createDefaultSnapshot());
+            const position = Vec3();
+            Vec3.scaleAndAdd(position, position, origin, 100);
+            plugin.canvas3d.camera.setState({ position }, 0);
+            const deltaDistance = Vec3();
+            Vec3.negate(deltaDistance, position);
+            if (Vec3.dot(deltaDistance, dirC) <= 0) {
+                Vec3.negate(plugin.canvas3d.camera.position, position);
+            }
+            const up = Vec3.create(0, 1, 0);
+            if (Vec3.dot(up, dirA) <= 0) {
+                Vec3.negate(plugin.canvas3d?.camera.up, plugin.canvas3d.camera.up);
+            }
+        }
+        const newPrincipalAxes = {
+            origin: origin,
+            dirA: dirA,
+            dirB: dirB,
+            dirC: dirC
+        };
+        return newPrincipalAxes;
+    }
+    return {
+        origin: Vec3.zero(),
+        dirA: Vec3.zero(),
+        dirB: Vec3.zero(),
+        dirC: Vec3.zero()
+    };
+}
+
+export function getPcaTransform(group: StructureComponentRef[]): [{principalAxes: PrincipalAxes}|undefined, Vec3|undefined] {
```

Comment:
Do you think it would be good idea to limit the size of the structures this is applied on? I can imagine for say >1M atoms (or even much lower), this method might bring little benefit and would actually be really slow?

Reply:
I think it is unnecessary since it depends on the size of polymer element of structure, which are usually much less than the atoms count

Reply:
Have you tried it on the HIV Cellpack model? (Load Cellpack in the left side of the plugin).

Reply:
I just tried 3j3k and 3j3y, the largest structure, and it works perfectly